Index Knuckles Controller Haptics

TheKrisSodroski opened the pull request:
This commit contains the base code to have haptics working. 

I've checked in `knuckles_right.json` and updated `actions.json` to include the required entries for haptics.  I'm unsure how actions.json is delivered to the end user, so I put the files in here for now as an example.  

Haptics are controlled through a `haptics.json` file, which defines the haptics for both single handed, and dual handed haptics for each weapon.  

I have not tweaked the values in this json file, but any help would be appreciated or suggestions would be appreciated.  

Code recognizes whether user is left handed or right handed, and can assign different values for dominant vs non-dominant vibrations.

97saundersj commented:
Wow great work, I know this is a highly requested feature and I can't wait to give this a test!
I had a quick look through the changes and its looking good, I like some of the refactoring changes, but I'll try and have a more in depth look at some point.
I'm just wondering why you mentioned only index knuckles controllers specifically, I'm guessing because that's the only controller you have access to for testing?
Also for the charged plasma pistol maybe you could check for how long the user has held the trigger button whilst holding the weapon? That would probably be too 'hacky' though.
Also it would be great to have the compiled zip files available for easy testing.

TheKrisSodroski commented:
> Wow great work, I know this is a highly requested feature and I can't wait to give this a test!

Thanks, but I can't stress enough that it's built on top of the shoulders of giants.  Your work is so good, that someone like myself without any vr or game experience was able to come in and quickly figure out how to add features.  This is unheard of in the real world! 👏👏👏

>I'm just wondering why you mentioned only index knuckles controllers specifically, I'm guessing because that's the only controller you have access to for testing?

Yes, I only have knuckles to test with.  

No, because I don't actually understand how the json files are being delivered to clients.

The `actions.json` file checked into the repo is different than the ones that's delivered to clients.  

![image](https://github.com/user-attachments/assets/271a025f-18ac-4970-b559-2617d0dd897d)

Not sure where these are coming from.  Do you manually make the release by copying these files (from somewhere other than the repo) and then zip it up to deliver to clients?  If so, maybe I can take a stab at building a release script to automatically do this. 

On top of that, in order to get haptics to work, I added the action to `knuckles_right.json`.  i assume by doing this, haptics won't actually work with other controllers until these entries are there.  I'm not a vr developer, so I'm not exactly sure how to add haptics to all controllers, or if there's a `generic` place to put these haptic actions.  

>Also for the charged plasma pistol maybe you could check for how long the user has held the trigger button whilst holding the weapon? That would probably be too 'hacky' though.

That was my additional thoughts.  The `haptics.json` file could probably define tick bound haptics so for things like the plasma pistol I can ramp up vibration as charge happens.  I'll start working towards refining that, but off the top of my head, I'm not sure how to recognize when the plasma pistol is actually fired (mouse release) so that I can trigger a large vibration.  Same with the "overheat" mechanism.  Maybe I can pick this off the weapon model somehow?  

>Also it would be great to have the compiled zip files available for easy testing.

If there's no distribution script, I can start to add that in so that haptics and other json files are automatically zipped up ready for client release.  We can probably built it into github actions to autorelease moving forward.  

Couple other notes: 

1.  I've gotten my fiancé to agree to help me flesh out the haptics values.  I'll check in a more refined `haptics.json` file next week when she has some time to modify the config while I test (as I added automatic haptics reloading on change).   
2. I'm wondering if you have some time in the coming weeks to maybe chat with me so I can ask some questions about how everything is working.  My next big goal would be to get multiplayer to work, but I'd like to mine your knowledge about how you decompiled and reverse engineered this.

LivingFray commented:
Nice work, I haven't had a chance to look over this properly yet but from what I've seen its looking great.
A few notes/comments:
- You should be able to detect the plasma pistol charging by detecting the player's shooting property remaining greater than zero while that gun is equipped (static_cast<UnitDynamicObject*>(Helpers::GetLocalPlayer())->Shooting)
- actions.json is delivered to the player by being copied into the /VR/OpenVR folder of HaloCEVR.zip by makerelease.bat and the mod instructs OpenVR to look there on startup
- I have a jank python script (manifest.py) for auto generating the binding files (including actions.json) which I manually run whenever I make changes to the default bindings, it should be fairly straightforward to add the haptics section to that to avoid having to manually update every json file. (That script also needs refactoring, it is quite the mess)
- You should probably go ahead and remove the scoped weapon enum and replace it with the more fleshed out one you've created

teddybear082 commented:
So in theory other controller users could copy the haptics part of the knuckles(right) json file to the bindings files for the other controllers it seems, if I understand this correctly?

TheKrisSodroski commented:
I’m gonna integrate with the python script early next week so that haptics works for all controllers.On Dec 8, 2024, at 6:22 AM, teddybear082 ***@***.***> wrote:﻿
So in theory other controller users could copy the haptics part of the knuckles(right) json file to the bindings files for the other controllers it seems, if I understand this correctly?

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you authored the thread.Message ID: ***@***.***>

TheKrisSodroski commented:
I've updated makerelease.bat to include haptics automatically. 

I've also added the vibration actions to the manifest.py file and refactored ScopedWeaponType to use WeaponType instead.

TheKrisSodroski commented:
> * You should be able to detect the plasma pistol charging by detecting the player's shooting property remaining greater than zero while that gun is equipped (static_cast<UnitDynamicObject*>(Helpers::GetLocalPlayer())->Shooting)

I gave this a shot, but it seems this location doesn't seem to correspond to the way we would need to deal with haptics for the plasma pistol. 

One issue, is that WeaponHandler.Prefire doesn't actually get called during charging, only when the weapon is actually fired.  For the plasma pistol, this is the "release" of the charge.  

To work around this, I had to hook into the main input hook and start the logic there.   Needs a lot of work, and is going to be hacky and error prone, but with some finesse it should provide a decent enough experience I think.

Branch for Plasma pistol: 

https://github.com/TheKrisSodroski/HaloCEVR/tree/plasmaPistol

97saundersj commented:
@TheKrisSodroski I want to give these changes a test but when I try and build the files I get these errors, do I need to do anything different from normal? 
![image](https://github.com/user-attachments/assets/badfc6df-0f7b-4a5f-a2b2-6def43febd69)
Is it worth trying to test your changes now or should we wait for your plasma pistol changes to be finished off first? I'm just excited to give them a try :)

TheKrisSodroski commented:
@97saundersj 

Oops.  Forgot to checkin the vs proj file with the new compiled files.  I've updated that so it should build correctly now. 

The work on haptics is basically done I think from my side.  I've implemented the plasma pistol as best as I could.  

Some issues: 

1.  There's no way for us to know if the player is reloading (as far as I can tell). 
2.  The different values for haptics don't seem to make much of a difference.  I've set the values to default 1 for frequency and amplitude for now.  It uses OpenVR TriggerHapticAction.  Not sure if TriggerHapticPulse would be better, but this seems to be on it's way to deprecation.  
3.  Unsure how to get this to work with vehicles. 

Other notes: 

1. Now that holsters are available, we probably should add haptics in the event that the user enters the holster area.

97saundersj commented:
Thanks for fixing that @TheKrisSodroski, I can build the files now but when actually launching Halo I get an error. I get a DirectX 9.0b needs to be installed error when launching halo, and the inject.log reads only 
```
WeaponHapticsConfig] Initializing
[WeaponHapticsConfig] Loading Config
``` 
So i'm not sure if it's crashing somewhere whilst trying to load the hapticsConfig? The haptics.json is placed in the "Halo/VR" folder.

With weapon holsters I'm happy to add haptics to them after this PR goes in,

I've had a look over your code and it looks good but just left a few minor comments.

97saundersj reviewed (changes requested):
A few minor comments, but getting issues with launching the game so I can't actually test the haptics.

TheKrisSodroski commented:
@97saundersj 

That issue should be resolved now.  Was missing a comma in the Bindings/haptics.json file.  

I'll start working on your feedback this weekend.  

Happy testing!

97saundersj commented:
I managed to test these changes and wow they feel great! The Assault Rifle in particular now feels like a proper gun and not a pea shooter!
I came across a few bugs while testing:
- When out of ammo the Plasma Pistol seemed to keep its overcharged vibrations
- If I fired with a weapon and then swapped to the Plasma Pistol it was vibrating like it was overcharging
- The needler didn't seem to have any vibrations
- When a weapon was out of ammo and held in two hands the foregrip was still vibrating after firing
- When fireing a weapon two handed it felt like the vibrations in the foregrip were stronger than the vibrations of the hand pulling the trigger (Not sure if this is actually the case but thats what it felt like to me, so might just be a personal thing)

Keep up the good work I'm very excited for these improved haptics :)

LivingFray commented:
> Some issues:
> 
>     1. There's no way for us to know if the player is reloading (as far as I can tell).

Try this, it should be some help
https://github.com/LivingFray/HaloCEVR/commit/3063953689de12ad29f7b9f56dea4d303bf37c43

Also, on an unrelated note, having separate json configs for each weapon got me thinking: I wonder if at some point in the future it would be worth expanding these files beyond haptics to make things more data-driven so that custom weapons can be better supported. 
For example having a json file that says any weapon with string "\\XXXX\\" in its filename should use these haptics, this scope offset, can't be two handed, etc.
Not something you should necessarily add in this PR, but something to think about perhaps.

teddybear082 commented:
Wow amazing work! Whenever this is ready for testing if someone can share the compiled version I am happy to

TheKrisSodroski commented:
@teddybear082 

I'll be implementing the review items this week.  Should have another version out later this week and I'll compile it up for you.

slipperfish commented:
> Wow amazing work! Whenever this is ready for testing if someone can share the compiled version I am happy to

I just happened to compile this and it works for me, let me know if it doesn't work or wait for Kris's update. 

[HaloCEVR.zip](https://github.com/user-attachments/files/18142868/HaloCEVR.zip)

teddybear082 commented:
Just reporting back - I tried the DLL shared by slipperfish above and I can feel the haptics but they are very low power (almost like the type of haptic you would get from picking up an item rather than firing a gun). On Quest 3 + VirtualDesktop. Would playing around with the "amplitude" value in the haptics.json work potentially to increase the haptics? Figured I would share either way.  Great work!

TheKrisSodroski commented:
> but they are very low power (almost like the type of haptic you would get from picking up an item rather than firing a gun)

Yes, I feel that as well. 

What's odd is that the amplitude is at it's maximum value according to the api documentation (between 0f and 1f, and we have 1f as the value).  

I might try and use "TriggerHapticPulse" instead, but I believe it's being deprecated.

LivingFray commented:
Have you tried adjusting the frequency? I believe the docs say it has a range of 0-360 and from a quick scan through the files you have it at 1 by default

LivingFray commented:
Hey, I've not really been keeping up with this PR. Beyond fixing the merge conflict from a recent change of mine, is this ready to merge into master? Or are there still some issues that need resolving first?

TheKrisSodroski commented:
With the holidays, it's been tough doing some last testing.  I'll be getting to it this weekend to make some final changes, but I think it should be good to go early next week.

LivingFray commented:
Alright, no worries. Just wanted to make sure I wasn't neglecting a completed feature
